### ISSUE TITLE
[Feature] Better way to load balance between testcases

### ISSUE THREAD

**Issue by alyaothman14** [microsoft/playwright#20116]
Let us know what functionality you'd like to see in Playwright and what your use case is.
Currently, we have some shards taking longer than others just because PW runs tests in alphabetical order, sometimes causing for some test cases to take longer than others. I would love to have load balancing for test cases. 

Ideas:
- Allow for the slowest test to run first across all available shards to run first by collecting data and producing a report that can be used in subsequent runs to get a better load balance.

Do you think others might benefit from this as well?
Faster execution time

**Comment by schlenks** [comment]
This is an idea similar to what [Knapsack](https://www.knapsackpro.com/) does for Cypress tests. It collects timings on all tests, creates a report and then uses timings on future test runs to load balance. If Playwright has a way to generate a test timings report and balance tests across shards from that report, it would be possible to consistently have optimized test run times. 

A few additional considerations:
- A report that estimates the impact on adjusting shards is really helpful. IE If you currently had 4 shards used, 3 shards +3 minutes overall time,  5 shards -1 minute overall time, 6 shards -3 minutes overall time, 7 shards -4 minutes overall time.
- This report would also be really helpful for highlighting what the slowest tests are in the test suite.

**Comment by unlikelyzero** [comment]
This is also part of the Cypress Dashboard Service

**Comment by dtinth** [comment]
Right now on one project with 800+ test files, we use [BullMQ](https://docs.bullmq.io/) with an externally-hosted Redis server to dynamically allocate tests to worker machines.

Instead of pre-allocating test tasks to workers at the beginning, we spawn a number of worker machines, and each of them **pulls** the tasks from the central queue. This helps ensure that resources are best utilized.

We also collect statistics how long each test file takes. We then use that info to re-order the tasks, so that slowest tests are pulled from the queue first. This way, towards the end only fast tests are remaining, and that helps even out the duration between each worker.

Having support for **pull queues** would really help in a large project.

**Comment by unlikelyzero** [comment]
@dtinth it would be great if you could share some of your code!

What you're describing is how the current Cypress Dashboard works.

**Comment by dtinth** [comment]
@unlikelyzero Most of the code is tightly coupled to our CI/CD setup, Redis instance, and test result database, and so can't be shared easily. But I can share the general approach. Fortunately most of the custom code is just glue code and most of the heavy lifting is done by BullMQ.

1. In GitHub Actions, I first determine the **queue ID** and then spawn 20x jobs in a matrix. Each job runs these scripts in order: The **enqueue jobs** script and **work on jobs** script. Afterwards, the **report** script runs at the end.

    - Every worker in the same workflow should use the same queue ID. Workers from different workflow should use a different queue ID.

2. The **enqueue jobs** script performs a glob of all test files and [adds them as a job to the queue](https://docs.bullmq.io/guide/queues/adding-bulks). Each filename is hashed to get a [job ID](https://docs.bullmq.io/guide/jobs/job-ids), which is used for job de-duplication. So, 20 workers can all try to add the same job to the queue, and thanks to deduplication, it will only be added once.

    - Jobs are dequeued in the same order that they are added. So the **test result database** can be queried to find out the duration of tests that run in the past. This duration can be used to sort the tests so that longer tests gets executed first and only short tests get run at the end, evening out the durations.

3. The **work on tasks** script launches a [BullMQ worker](https://docs.bullmq.io/guide/workers). The worker just runs the test file and reports the status back to BullMQ. Once the `'drained'` event is emitted, it means that there are no more tests to run, and the worker can exit.

    - There are several open source UI for BullMQ, so you can deploy a BullMQ dashboard and use it to track its progress in real-time.
    - The test report for each job is written to different directories and uploaded to an artifact storage. If you use GitHub Actions, uploading multiple small files will be smaller than uploading one big file. So in our setup, we zip up all the generated artifacts and upload them after the worker has finished processing testing jobs.

4. After all tests are finished, the **report** script will query all jobs from the queue (BullMQ persists all jobs until it is [deleted](https://docs.bullmq.io/guide/queues/removing-jobs)) and generate a report, set the CI status, and update the test results database with their durations. It also downloads all the uploaded artifacts.

    - It is important that the report contains the information that can identify which worker ran which test. Having that info will become very helpful for debugging when tests failed.

There is an overhead of spawning a test runner for each file, but that is outweighed by getting the ability to run 20 workers simultaneously.

**Comment by emperol2** [comment]
Thanks @dtinth !

**Comment by liamdebeasi** [comment]
Ionic Framework's CI process would benefit greatly from this feature. Many of our tests rely on screenshot comparisons or simulating user interaction, so the runtime adds up. We use the sharding feature to distribute tests over `n` test runners which helps, but the run time of each test runner is uneven. As a result, some test runners take 3 minutes and other can take 13 minutes. Having a way of ensuring that each test runner takes (roughly) the same amount of time would noticeably reduce our CI run times.

**Comment by agoldis** [comment]
We did something similar and could reduce the overall duration by -30%. Indeed there's an an overhead of spawning a test runner for each file, but if your testing suite has many spec files, the overall benefit is quite good!

https://currents.dev/readme/guides/pw-parallelization/playwright-orchestration

**Comment by rohitkrishna-marneni** [comment]
I see that Issue [#27356](https://github.com/microsoft/playwright/issues/27356) is closed linking to this issue. But I still have rudimentary problem where playwright is unable to split tests into shards evenly. Like I have 100 tests and 10 shards, I would expect 10 tests to be split to each shard. However 4-5 shards are exiting immediately and the remaining shards are getting like 20 tests each. I feel like this would be easier to fix before thinking about load balancing using more complex approaches(via test timings and what not).

**Comment by muhqu** [comment]
I like to revive discussion on how to improve playwright sharding using timing data: https://github.com/microsoft/playwright/issues/17969#issuecomment-2421653419

**Comment by dtinth** [comment]
Just an update, now we are using this setup to run tests in parallel:

- [**Microsoft Playwright Testing**](https://azure.microsoft.com/en-us/products/playwright-testing), a Microsoft Azure paid service. Unlike other paid services that just sell a test result dashboard and test orchestrator, Microsoft Playwright Testing service lets you rent a browser on the cloud, and automatically handles tunneling (traffic from the browser will be (optionally) tunneled through the machine that runs `playwright test`, letting it access localhost and other services securely). So you can run a single application server instance, and launch 50 browsers on the cloud to test it in parallel [with a single configuration change that can be automated in a single command](https://learn.microsoft.com/en-us/azure/playwright-testing/quickstart-run-end-to-end-tests?tabs=playwrightcli#install-microsoft-playwright-testing-package).

- [**Larger runners on GitHub Actions.**](https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners) As the heavy lifting of running browsers is now offloaded to Azure, now the bottleneck becomes the application server under test becoming unable to handle the traffic from 50 browsers. This caused every single test to take longer, consuming more test minutes. To fix this, we upgraded to a larger runner so that app server can handle more traffic. Since we run just one app server, resource is better utilized. Since each tests now run faster, the number of test minutes significantly reduced, offsetting the cost of upgrading the runner. We have to experiment with multiple runner sizes and browsers to find out which configuration is the most cost effective.

tl;dr: we tried these paid microsoft services and they are great

**Comment by ValorHeart** [comment]
@dtinth How is that related to load balancing between short and long tests between shards?

**Comment by dtinth** [comment]
@ValorHeart Well, we no longer have to shard, because the heavy lifting of running browsers has been outsourced to external service, we can now run tests with a single shard but multiple workers (which automatically load-balances between short and long tests).

**Comment by szymonnowak-st** [comment]
@pavelfeldman Any chance this could be reconsidered? I know sharding was added recently, but it doesn't quite cover this use case.

We're currently hacking `Dispatcher#run` to run the slowest tests first, which gives us a signifcant boost in total runtime. We have to sometimes update our hack when we upgrade Playwright, but it's still worth it.

**before**
<img width="400" alt="Image" src="https://github.com/user-attachments/assets/dc99a1b1-e990-4ca5-8618-6f8c7c41a83f" />

**after**
<img width="400" alt="Image" src="https://github.com/user-attachments/assets/62bd8e9d-a331-4597-a4f3-a7244b5d0d7b" />

**Comment by lukpsaxo** [comment]
shouldn't this be a enhancement request generally to change the ordering in all scenarios?
